File: src/main/java/io/fabric8/launcher/booster/catalog/rhoar/BoosterPredicates.java
package io.fabric8.launcher.booster.catalog.rhoar;

import java.util.List;
import java.util.Map;
import java.util.function.Predicate;

import javax.annotation.Nullable;

import io.fabric8.launcher.booster.catalog.Booster;
import io.fabric8.launcher.booster.catalog.rhoar.predicates.BoosterParameterPredicate;
import io.fabric8.launcher.booster.catalog.rhoar.predicates.BoosterScriptingPredicate;

/**
 * Utility class to provide {@link Predicate} instances for {@link Booster} instances and its subtypes
 */
public final class BoosterPredicates {
    private BoosterPredicates() {
        throw new IllegalAccessError("Utility class");
    }

    /**
     * Returns a {@link Predicate} for a {@link RhoarBooster} testing if its {@link Runtime}
     * is equal to the provided parameter
     *
     * @param runtime the {@link Runtime} to be compared against a given {@link RhoarBooster}.
     *                If null, always returns true;
     * @return a {@link Predicate} testing against the given {@link Runtime}
     */
    public static Predicate<RhoarBooster> withRuntime(@Nullable Runtime runtime) {
        return (RhoarBooster b) -> runtime == null || runtime.equals(b.getRuntime());
    }

    /**
     * Returns a {@link Predicate} for a {@link RhoarBooster} testing if its {@link Mission}
     * is equal to the provided parameter
     *
     * @param mission the {@link Mission} to be compared against a given {@link RhoarBooster}.
     *                If null, always returns true;
     * @return a {@link Predicate} testing against the given {@link Mission}
     */
    public static Predicate<RhoarBooster> withMission(@Nullable Mission mission) {
        return (RhoarBooster b) -> mission == null || mission.equals(b.getMission());
    }

    /**
     * Returns a {@link Predicate} for a {@link RhoarBooster} testing if its {@link Version}
     * is equal to the provided parameter
     *
     * @param version the {@link Version} to be compared against a given {@link RhoarBooster}.
     *                If null, always returns true;
     * @return a {@link Predicate} testing against the given {@link Version}
     */
    public static Predicate<RhoarBooster> withVersion(@Nullable Version version) {
        return (RhoarBooster b) -> version == null || version.equals(b.getVersion());
    }

    /**
     * Returns a {@link Predicate} for a {@link RhoarBooster} testing if the runsOn attribute defined inside
     * its metadata matches the provided parameter
     *
     * @param clusterType the cluster type to be compared against a given {@link RhoarBooster}.
     *                    If null, always returns true;
     * @return a {@link Predicate} testing against the given cluster type
     * @see RhoarBooster#checkCategory(List, String)
     */
    public static Predicate<RhoarBooster> withRunsOn(@Nullable String clusterType) {
        return (RhoarBooster b) -> b.runsOn(clusterType);
    }

    /**
     * Returns a {@link Predicate} for a {@link Booster} testing if against a script expression that must be evaluated
     *
     * @param script the script expression to be tested against a given {@link Booster}. Should never be null.
     * @return a {@link BoosterScriptingPredicate} instance
     */
    public static <T extends Booster> Predicate<T> withScriptFilter(@Nullable String script) {
        //noinspection unchecked
        return (script != null) ? (Predicate<T>) new BoosterScriptingPredicate(script) : b -> true;
    }

    /**
     * Returns a {@link Predicate} for a {@link Booster} testing if the given parameters match.
     *
     * @param parameters The parameters belonging to the tested {@link Booster} instance. Should never be null
     * @return a {@link BoosterScriptingPredicate} instance
     * @throws NullPointerException if parameters is null
     */
    public static <T extends Booster> Predicate<T> withParameters(Map<String, List<String>> parameters) {
        //noinspection unchecked
        return (Predicate<T>) new BoosterParameterPredicate(parameters);
    }

    /**
     * Returns a {@link Predicate} for a {@link Booster} testing if the given application argument
     * matches a property in the metadata section named "app.$application.enabled". The booster will
     * be filtered out if the property evaluated to "false".
     *
     * @param application The name of the application
     * @return a {@link Predicate} testing if ttthe given application is enabled
     */
    public static Predicate<RhoarBooster> withAppEnabled(@Nullable String application) {
        return (RhoarBooster b) -> application == null || b.getMetadata("app/" + application + "/enabled", true);
    }
}

File: src/main/kotlin/io/fabric8/launcher/booster/catalog/rhoar/BoosterPredicates.kt
package io.fabric8.launcher.booster.catalog.rhoar

import java.util.function.Predicate

import io.fabric8.launcher.booster.catalog.Booster
import io.fabric8.launcher.booster.catalog.rhoar.predicates.BoosterParameterPredicate
import io.fabric8.launcher.booster.catalog.rhoar.predicates.BoosterScriptingPredicate
import java.util.regex.Pattern

/**
 * Utility class to provide [Predicate] instances for [Booster] instances and its subtypes
 */
class BoosterPredicates private constructor() {
    init {
        throw IllegalAccessError("Utility class")
    }

    companion object {

        /**
         * Returns a [Predicate] for a [RhoarBooster] testing if its [Runtime]
         * is equal to the provided parameter
         *
         * @param runtime the [Runtime] to be compared against a given [RhoarBooster].
         * If null, always returns true;
         * @return a [Predicate] testing against the given [Runtime]
         */
        @JvmStatic
        fun withRuntime(runtime: Runtime?) = Predicate<RhoarBooster> { b: RhoarBooster ->
            runtime == null || runtime == b.runtime
        }

        /**
         * Returns a [Predicate] for a [RhoarBooster] testing if the provided [Pattern]
         * matches the booster's [Runtime.id]
         *
         * @param version the [Pattern] to be compared against a given [RhoarBooster] version.
         * If null, always returns true;
         * @return a [Predicate] testing against the given [Version]
         */
        @JvmStatic
        fun withRuntimeMatches(runtime: Pattern?) = Predicate<RhoarBooster> { b: RhoarBooster ->
            runtime == null || b.runtime == null || runtime.matcher(b.runtime?.id).matches()
        }

        /**
         * Returns a [Predicate] for a [RhoarBooster] testing if its [Mission]
         * is equal to the provided parameter
         *
         * @param mission the [Mission] to be compared against a given [RhoarBooster].
         * If null, always returns true;
         * @return a [Predicate] testing against the given [Mission]
         */
        @JvmStatic
        fun withMission(mission: Mission?) = Predicate<RhoarBooster> { b: RhoarBooster ->
            mission == null || mission == b.mission
        }

        /**
         * Returns a [Predicate] for a [RhoarBooster] testing if its [Version]
         * is equal to the provided parameter
         *
         * @param version the [Version] to be compared against a given [RhoarBooster].
         * If null, always returns true;
         * @return a [Predicate] testing against the given [Version]
         */
        @JvmStatic
        fun withVersion(version: Version?) = Predicate<RhoarBooster> { b: RhoarBooster ->
            version == null || version == b.version
        }

        /**
         * Returns a [Predicate] for a [RhoarBooster] testing if its [Pattern]
         * matches the booster's [Version.id]
         *
         * @param version the [Pattern] to be compared against a given [RhoarBooster] version.
         * If null, always returns true;
         * @return a [Predicate] testing against the given [Version]
         */
        @JvmStatic
        fun withVersionMatches(version: Pattern?) = Predicate<RhoarBooster> { b: RhoarBooster ->
            version == null || b.version == null || version.matcher(b.version?.id).matches()
        }

        /**
         * Returns a [Predicate] for a [RhoarBooster] testing if the runsOn attribute defined inside
         * its metadata matches the provided parameter
         *
         * @param clusterType the cluster type to be compared against a given [RhoarBooster].
         * If null, always returns true;
         * @return a [Predicate] testing against the given cluster type
         * @see RhoarBooster.checkCategory
         */
        @JvmStatic
        fun withRunsOn(clusterType: String?) = Predicate<RhoarBooster> { b: RhoarBooster -> b.runsOn(clusterType) }

        /**
         * Returns a [Predicate] for a [Booster] testing if against a script expression that must be evaluated
         *
         * @param script the script expression to be tested against a given [Booster].
         * @return a [BoosterScriptingPredicate] instance
         */
        @JvmStatic
        fun <T : Booster> withScriptFilter(script: String?) =
                if (script != null)
                    BoosterScriptingPredicate(script) as Predicate<T>
                else
                    Predicate<T> { _ -> true }

        /**
         * Returns a [Predicate] for a [Booster] testing if the given parameters match.
         *
         * @param parameters The parameters belonging to the tested [Booster] instance. Should never be null
         * @return a [BoosterScriptingPredicate] instance
         * @throws NullPointerException if parameters is null
         */
        @JvmStatic
        fun <T : Booster> withParameters(parameters: Map<String, List<String>>) =
                BoosterParameterPredicate(parameters) as Predicate<T>

        /**
         * Returns a [Predicate] for a [Booster] testing if the given application argument
         * matches a property in the metadata section named "app.$application.enabled". The booster will
         * be filtered out if the property evaluated to "false".
         *
         * @param application The name of the application
         * @return a [Predicate] testing if ttthe given application is enabled
         */
        @JvmStatic
        fun withAppEnabled(application: String?) = Predicate<RhoarBooster> { b: RhoarBooster ->
            application == null || b.getMetadata("app/${application.toLowerCase()}/enabled", true)
        }
    }
}
